rfcs/0121-hardware-driver.md

```diff
@@ -0,0 +1,88 @@
+The usage of opengl explicitly may have been an accurate name given the time
+in which NixOS was first created; however, graphic drivers alone include much
+more than just userland graphics libraries so this usages is now misaligned with
+current paradigms. The misalignment with this path and the related Nixpkgs
+utilities are a misnomer in most contexts, and contributes to Nixpkgs' 
+"weirdness budget".
+
+# Motivation
+[motivation]: #motivation
+
+Video drivers ship many hardware acceleration libraries these days: vulkan,
+opengl, opencl, video encoding, cuda, and many other userland libraries are
+installed. Having all of these libraries placed under a single "opengl" header
+is odd for many contributors, especially newcomers. This awkwardness is
+compounded with other specialized hardware such as fpgas, tpus, and asics
+where their function isn't related to graphics, however, also need to make
+use of a "known good" path in which the related userland libraries will be mounted.
+
+To remedy this misalignment, a new convention around placing those libraries
+should be used, preferably `/run/current-system/drivers/`.
+
+# Detailed design
+[design]: #detailed-design
+
+Most of the implementation of the current paradigms will continue, as there
+will always be a need for some "impurity" around hardware based libraries.
+The significant changes will be around naming conventions used within
+Nixpkgs and NixOS.
+
+Deprecate and move existing `/run/opengl-driver/` logic:
+- Rename `hardware.opengl` options to `hardware.drivers` (pre 22.11)
+- Rename `pkgs.addOpenGLRunpath` shell hook to `addHardwareRunpath` (pre 22.11)
+  - Alias `addOpenGLRunpath` to `addHardwareRunpath` for compatibility (pre 22.11)
+- Update nixpkgs references of `/run/opengl-driver/` to point to `/run/current-system/drivers/` (post 22.11)
+- Update `mesa.driverLink` to point to `/run/current-system/drivers/lib` (post 22.11)
+
+For compatibility with existing nixpkgs packages, `/run/opengl-driver{,-32}/` will
+be a symbolic link to `/run/current-system/drivers{,-32}/`. This will likely be
+an indefinite change, or else older packages will not work on NixOS. Also,
+we need to ensure compatibility with out-of-tree code which may have been built around
+the opengl paths, such as [nixGL](https://github.com/guibou/nixGL).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Technical churn. Doesn't provide any new or added benefit, other than
+correcting a stale misnomer.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Continue to use `/run/opengl-driver` in it's current state.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Is there a better name for the path, nixos module options, and nixpkgs hook?
+
+This is also an opportunity to revisit the `hardware.opengl` module. Some
+potential improvements could be:
+- Move `hardware.opengl.package` + `hardware.opengl.extraPackages` to a single `hardware.drivers.packages`
```

Comment:
This makes replacing mesa in particular harder, as now it's a single list. Using an overlay isn't practical as it rebuilds far too much.

Making this a set so you can do `hardware.drivers.packages.mesa = myCustomMesa` could solve that.  Leaving it as is after the rename also works.

I originally commented this on the impl PR instead of the RFC accidentally.

Reply:
How special is `mesa`? Does everything use it now?

I will concede that if it is special, that is not just an OpenGL thing, but also something that could be true in a Vulkan / many languages world.

Reply:
Having it as a set makes it easier to replace any specific package, we'd change anywhere in nixpkgs that sets it to eg do `hardware.drivers.packages.amdvlk-pro = something` or `hardware.drivers.packages.nvidia` so it wouldn't be mesa specific and makes overriding any specific package easier.

Mesa is sort of special - any system with working gl/vulkan is expected to have it, even when using other proprietary libs alongside it - but I don't think handling specifically for mesa is the right approach when we could make it easier to override any `hardware.drivers` package with the right approach.

Reply:
I agree with the general point here that in the modules system, lists are poorly composable. They can only be appended to, or replaced outright. We don't have the tools to allow a user to remove or replace an element from a configuration system list.

There are too many options already that imo should be migrated to attrsets to let the user be in control. Let's not add more if we can.

Reply:
I wouldn't set across the board because names are arbitrary. Like, if I do `foo = <Nvidia-drivers-package>:` whose to say I'm doing anything wrong?

Reply:
You aren't doing anything wrong if you do that and it isn't much of a problem if you do that, unless you get it included in nixpkgs I guess.

If you do that alongside something else setting `nvidia = <another nvidia drivers package>` you could get two version of the same package by mistake. That could be prevented by asserting that the pnames of the packages involved are unique although I'm not sure that's necessary.

Reply:
I am not sure how express this, but we shouldn't design things around names. Perhaps other modules can make configurable what package they would add to this list, but unless we can characterize "structurally" how these package fit into different roles I would leave it just as is.

Reply:
Using a list to avoid the names is even worse, as your only options are to use an overlay and rebuild everything or to mkForce and have to keep track of any other entries that should be in the list.

If we keep it as is and have .package and .extraPackages then we aren't relying on names but we're also only making it easy to override mesa, and it's even easier to accidentally include duplicate packages of different versions in the list as instead of relying on a name there's nothing to prevent dupes.